### PR TITLE
Added support for min and max month

### PR DESCRIPTION
**Pull request by kevinvanmierlo**
Added support for min and max month. This is the way the pager will now behave, so let me know if that's okay. It calculates how many pages there should be in LazyRow based on minMonth and maxMonth. Default it uses 10000 months in the past and 10000 in the future so it works the same as it already did. This fixes https://github.com/boguszpawlowski/ComposeCalendar/issues/112.

### CODE REVIEW

**Comment by kevinvanmierlo** [comment]
@boguszpawlowski Do I need to do anything? It says Review required, but I'm assuming that is a review by you right?

**Review by boguszpawlowski — COMMENTED** [review]
@kevinvanmierlo Thank you for the contribution, and apologies for the CR wait time. Overall, implementation LGTM, couple of minor comments:

**Comment by kevinvanmierlo** [comment]
@boguszpawlowski No problem! Thanks for responding and adding the test results. I created a new commit which should fix the issues. I'll also reply on each test result to let you know what I've done.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/header/MonthState.kt [hunk 1]
```diff
@@ -5,34 +5,70 @@ import androidx.compose.runtime.getValue
...
     @Suppress("FunctionName") // Factory function
     public fun Saver(): Saver<MonthState, String> = Saver(
-      save = { it.currentMonth.toString() },
-      restore = { MonthState(YearMonth.parse(it)) }
+      save = { monthState ->
+        JSONObject().also { jsonObject ->
+          jsonObject.put("currentMonth", monthState.currentMonth.toString())
+          jsonObject.put("minMonth", monthState.minMonth.toString())
+          jsonObject.put("maxMonth", monthState.maxMonth.toString())
+        }.toString()
```
[the remark this answers is not in the document]
Comment (kevinvanmierlo) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/header/MonthState.kt]: Thanks! Don't really use savers a lot, so I didn't know about those savers.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/month/MonthContent.kt [hunk 2]
```diff
@@ -77,13 +79,16 @@ internal fun <T : SelectionState> MonthPager(
...
+      ChronoUnit.MONTHS.between(monthState.minMonth, monthState.maxMonth).toInt() + 1
+    }
     LazyRow(
       modifier = modifier.testTag("MonthPager"),
       state = listState,
       flingBehavior = flingBehavior,
       verticalAlignment = Alignment.Top,
     ) {
-      items(PagerItemCount) { index ->
+      items(pagerCount, key = { monthListState.getMonthForPage(it).let { "${it.year}-${it.monthValue}" } }) { index ->
```
[the remark this answers is not in the document]
Comment (kevinvanmierlo) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/month/MonthContent.kt]: I agree, I also adjusted the code for better readability

File: sample/src/main/java/io/github/boguszpawlowski/composecalendar/sample/MinMaxCalendarSample.kt [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    StaticCalendar(calendarState = calendarState)
+
+    MinMaxControls(calendarState = calendarState)
+  }
+}
+
+@Composable
+private fun MinMaxControls(
+  calendarState: CalendarState<EmptySelectionState>,
+) {
```
[the remark this answers is not in the document]
Comment (kevinvanmierlo) [reply sample/src/main/java/io/github/boguszpawlowski/composecalendar/sample/MinMaxCalendarSample.kt:56]: I agree that would be better. I changed it to the way you explained.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/Calendar.kt [hunk 4]
```diff
@@ -256,7 +262,13 @@ public fun rememberSelectableCalendarState(
 @Composable
 public fun rememberCalendarState(
   initialMonth: YearMonth = YearMonth.now(),
+  minMonth: YearMonth = initialMonth.minusMonths(10000),
+  maxMonth: YearMonth = initialMonth.plusMonths(10000),
```
[the remark this answers is not in the document]
Comment (kevinvanmierlo) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/Calendar.kt]: Fixed it

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/Calendar.kt [hunk 5]
```diff
@@ -235,11 +235,17 @@ public fun <T : SelectionState> Calendar(
...
   initialSelection: List<LocalDate> = emptyList(),
   initialSelectionMode: SelectionMode = SelectionMode.Single,
   confirmSelectionChange: (newValue: List<LocalDate>) -> Boolean = { true },
   monthState: MonthState = rememberSaveable(saver = MonthState.Saver()) {
-    MonthState(initialMonth = initialMonth)
+    MonthState(
+      initialMonth = initialMonth,
+      minMonth = minMonth,
+      maxMonth = maxMonth
+    )
```
[the remark this answers is not in the document]
Comment (kevinvanmierlo) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/Calendar.kt]: While I agree this should be done for feature parity, I think this is out of the scope for this pull request. This pull request is to fix the min and max month for swiping and the WeekCalendar doesn't use the MonthYear to my knowing. The same principle could be applied to the WeekCalendar. But I would create a separate issue for that.

[the remark this answers is not in the document]
Comment (boguszpawlowski) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/Calendar.kt]: Agreed, I will create the ticket.

File: sample/src/main/java/io/github/boguszpawlowski/composecalendar/sample/CustomSelectionSample.kt [hunk 6]
```diff
@@ -47,9 +47,15 @@ private class MonthSelectionState(
 @Composable
 private fun rememberMonthSelectionState(
   initialMonth: YearMonth = YearMonth.now(),
+  minMonth: YearMonth = initialMonth.minusMonths(10000),
+  maxMonth: YearMonth = initialMonth.plusMonths(10000),
```
Comment (boguszpawlowski) [inline sample/src/main/java/io/github/boguszpawlowski/composecalendar/sample/CustomSelectionSample.kt:51]: Could reuse DefaultCalendarMonthRange constant.

Comment (kevinvanmierlo) [reply sample/src/main/java/io/github/boguszpawlowski/composecalendar/sample/CustomSelectionSample.kt:51]: @boguszpawlowski I don't mind reusing it, but that would mean to make it a public const. I didn't see any public const in the current library, so I chose to make it internal like the rest of the code. 

Do you want to make it public? I'll await your response before changing any code.

**Comment by boguszpawlowski** [comment]
@kevinvanmierlo Looks good to go. Please adjust errors reported by Detekt (https://github.com/boguszpawlowski/ComposeCalendar/actions/runs/7540813612/job/20526357349?pr=113) and re-generate api scheme `./gradlew apiDump`

**Review by boguszpawlowski — APPROVED** [review]

**Comment by kevinvanmierlo** [comment]
@boguszpawlowski I fixed the detekt issues. But it says it is still waiting for status to be reported. Is this something you need to run? I always thought it was automatic.

**Comment by boguszpawlowski** [comment]
@kevinvanmierlo Thanks again for the contribution! Will release the new version of the library soon.

**Comment by kevinvanmierlo** [comment]
No problem! It's a nice library :) 
Do you have an estimate on when it'll be released? I'm using my own fork for now, but rather use this version.

**Comment by boguszpawlowski** [comment]
Probably sometime next week. I want to add the min/max for week calendar, bump libraries and test it first.

**Comment by kevinvanmierlo** [comment]
Thank you! I get that you want to add that first. I'm looking forward to the release!

**Comment by boguszpawlowski** [comment]
@kevinvanmierlo Released  1.2.0. Should be available on Maven soon.

**Pull request merged**